File: articles/genes-inheritance-principles-of-heredity/python/chi_square_inheritance.py
"""
Chi-square inheritance tests for Mendelian ratios.

Run:
    python python/chi_square_inheritance.py
"""

from __future__ import annotations

from pathlib import Path

import pandas as pd


ARTICLE_DIR = Path(__file__).resolve().parents[1]
MONO_PATH = ARTICLE_DIR / "data" / "monohybrid_counts.csv"
DI_PATH = ARTICLE_DIR / "data" / "dihybrid_counts.csv"


def chi_square_from_ratio(df: pd.DataFrame) -> tuple[pd.DataFrame, float]:
    """Calculate expected counts and chi-square statistic from expected ratios."""

    total = df["observed"].sum()
    out = df.copy()
    out["expected"] = total * out["expected_ratio"]
    out["chi_component"] = (out["observed"] - out["expected"]) ** 2 / out["expected"]
    chi_square = float(out["chi_component"].sum())
    return out, chi_square


def main() -> None:
    """Run monohybrid and dihybrid chi-square examples."""

    mono = pd.read_csv(MONO_PATH)
    di = pd.read_csv(DI_PATH)

    mono_result, mono_chi = chi_square_from_ratio(mono)
    di_result, di_chi = chi_square_from_ratio(di)

    print("Monohybrid 3:1 test")
    print(mono_result.round(4).to_string(index=False))
    print("chi_square:", round(mono_chi, 4))

    print("\nDihybrid 9:3:3:1 test")
    print(di_result.round(4).to_string(index=False))
    print("chi_square:", round(di_chi, 4))


if __name__ == "__main__":
    main()

File: articles/genes-inheritance-principles-of-heredity/r/chi_square_inheritance.R
# Chi-square goodness-of-fit test for Mendelian ratios.

library(dplyr)

mono_path <- file.path("data", "monohybrid_counts.csv")
di_path <- file.path("data", "dihybrid_counts.csv")

if (!file.exists(mono_path)) {
  mono_path <- file.path("..", "data", "monohybrid_counts.csv")
  di_path <- file.path("..", "data", "dihybrid_counts.csv")
}

run_chi_square <- function(df) {
  total <- sum(df$observed)

  df %>%
    mutate(
      expected = total * expected_ratio,
      chi_component = (observed - expected)^2 / expected
    )
}

mono <- read.csv(mono_path)
di <- read.csv(di_path)

mono_result <- run_chi_square(mono)
di_result <- run_chi_square(di)

print(mono_result)
cat("Monohybrid chi-square =", round(sum(mono_result$chi_component), 4), "\n")

print(di_result)
cat("Dihybrid chi-square =", round(sum(di_result$chi_component), 4), "\n")
